File: Sources/Primitives/Signer.swift
import Foundation
import CryptoKit
// filename: Signer.swift
//
// Ed25519 message signing — ADR D12 / signing.md §1.
//
// Private keys live in the macOS Keychain (generic password, service
// `org.spy.yx.sig`, account = <agent-id>, value = base64 of the 32-byte
// seed). Mirrors the MeshKey `security`-subprocess pattern so the Swift and
// Python implementations share the exact same store/format. Public key =
// base64 of the 32 raw bytes; signatures are "ed25519:<base64 64-byte sig>".
//
// NOTE: CryptoKit's Ed25519 signing is randomized (RFC 8032-verifiable, but
// signing the same bytes twice yields different — equally valid — signatures).
// Fixture stability therefore rests on the canonical bytes + verification,
// not on byte-identical re-signing.

public enum Signer {

    public static let service = "org.spy.yx.sig"
    public static let sigPrefix = "ed25519:"

    // MARK: - Key management (Keychain-backed)

    /// Generate a keypair for `agentID`, store the private seed in the
    /// Keychain, and return the public key (base64, 32 raw bytes).
    public static func generate(agentID: String) -> String? {
        let key = Curve25519.Signing.PrivateKey()
        let seedB64 = key.rawRepresentation.base64EncodedString()
        guard keychainSet(agentID: agentID, seedB64: seedB64) else { return nil }
        return key.publicKey.rawRepresentation.base64EncodedString()
    }

    /// The stored public key for `agentID` (base64), or nil if absent.
    public static func publicKey(agentID: String) -> String? {
        privateKey(agentID: agentID)?.publicKey.rawRepresentation.base64EncodedString()
    }

    /// Agent-ids with signing keys present in the Keychain (sorted).
    public static func list() -> [String] {
        guard let dump = runSecurity(["dump-keychain"]) else { return [] }
        // Attributes are alphabetical, so "acct" precedes "svce" within an item.
        func quoted(_ s: Substring) -> String? {
            guard let r = s.range(of: "=\"", options: .backwards), s.hasSuffix("\"") else { return nil }
            return String(s[r.upperBound...].dropLast())
        }
        var ids = Set<String>(); var pendingAcct: String? = nil
        for line in dump.split(separator: "\n") {
            if line.contains("\"acct\"<blob>=") { pendingAcct = quoted(line) }
            if line.contains("\"svce\"<blob>=\"\(service)\""), let a = pendingAcct {
                ids.insert(a)
            }
        }
        return ids.sorted()
    }

    /// Delete the signing key for `agentID`.
    @discardableResult
    public static func remove(agentID: String) -> Bool {
        runSecurity(["delete-generic-password", "-s", service, "-a", agentID]) != nil
    }

    // MARK: - Sign / verify

    /// Sign `bytes` with the Keychain key for `agentID` → "ed25519:<base64>".
    public static func sign(_ bytes: Data, agentID: String) -> String? {
        privateKey(agentID: agentID).flatMap { sign(bytes, privateKey: $0) }
    }

    /// Sign `bytes` with an explicit 32-byte seed (base64) — tests/fixtures.
    public static func sign(_ bytes: Data, seedB64: String) -> String? {
        privateKey(seedB64: seedB64).flatMap { sign(bytes, privateKey: $0) }
    }

    /// Public key (base64) derived from an explicit seed — tests/fixtures.
    public static func publicKey(seedB64: String) -> String? {
        privateKey(seedB64: seedB64)?.publicKey.rawRepresentation.base64EncodedString()
    }

    /// Verify an "ed25519:<base64>" signature over `bytes` with a base64 pubkey.
    public static func verify(_ bytes: Data, sig: String, pubkeyB64: String) -> Bool {
        guard sig.hasPrefix(sigPrefix),
              let sigData = Data(base64Encoded: String(sig.dropFirst(sigPrefix.count))),
              sigData.count == 64,
              let pubData = Data(base64Encoded: pubkeyB64), pubData.count == 32,
              let pub = try? Curve25519.Signing.PublicKey(rawRepresentation: pubData)
        else { return false }
        return pub.isValidSignature(sigData, for: bytes)
    }

    // MARK: - Internals

    private static func sign(_ bytes: Data, privateKey: Curve25519.Signing.PrivateKey) -> String? {
        guard let sig = try? privateKey.signature(for: bytes) else { return nil }
        return sigPrefix + sig.base64EncodedString()
    }

    private static func privateKey(seedB64: String) -> Curve25519.Signing.PrivateKey? {
        guard let seed = Data(base64Encoded: seedB64), seed.count == 32 else { return nil }
        return try? Curve25519.Signing.PrivateKey(rawRepresentation: seed)
    }

    private static func privateKey(agentID: String) -> Curve25519.Signing.PrivateKey? {
        guard let b64 = keychainGet(agentID: agentID) else { return nil }
        return privateKey(seedB64: b64)
    }

    // MARK: - Keychain (generic password via `security`, mirroring MeshKey)

    static func keychainGet(agentID: String) -> String? {
        let out = runSecurity(["find-generic-password", "-s", service, "-a", agentID, "-w"])
        guard let b64 = out?.trimmingCharacters(in: .whitespacesAndNewlines), !b64.isEmpty else { return nil }
        return b64
    }

    @discardableResult
    static func keychainSet(agentID: String, seedB64: String) -> Bool {
        // -U upserts (update if present)
        runSecurity(["add-generic-password", "-s", service, "-a", agentID,
                     "-w", seedB64, "-U", "-l", "yx-sig-\(agentID)"]) != nil
    }

    private static func runSecurity(_ args: [String]) -> String? {
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/usr/bin/security")
        p.arguments = args
        let pipe = Pipe(); p.standardOutput = pipe; p.standardError = Pipe()
        do { try p.run() } catch { return nil }
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        p.waitUntilExit()
        guard p.terminationStatus == 0 else { return nil }
        return String(data: data, encoding: .utf8) ?? ""
    }
}

File: src/python/yx/primitives/signer.py
"""Ed25519 message signing — ADR D12 / signing.md §1.

Python mirror of Sources/Primitives/Signer.swift.

Private keys live in the macOS Keychain (generic password, service
`org.spy.yx.sig`, account = <agent-id>, value = base64 of the 32-byte seed).
Mirrors the MeshKey `security`-subprocess pattern so the Swift and Python
implementations share the exact same store/format. Public key = base64 of
the 32 raw bytes; signatures are "ed25519:<base64 64-byte sig>".

NOTE: `cryptography`'s Ed25519 signing is deterministic (RFC 8032); CryptoKit's
is randomized. Both verify each other — the cross-language contract is VERIFY,
not signature byte-equality (see tests/fixtures/sign-vector.sxp).
"""

from __future__ import annotations

import base64
import subprocess
from typing import List, Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.exceptions import InvalidSignature

SERVICE = "org.spy.yx.sig"
SIG_PREFIX = "ed25519:"


# MARK: - Key management (Keychain-backed)

def generate(agent_id: str) -> Optional[str]:
    """Generate a keypair for `agent_id`, store the private seed in the
    Keychain, and return the public key (base64, 32 raw bytes)."""
    key = Ed25519PrivateKey.generate()
    seed_b64 = base64.b64encode(key.private_bytes_raw()).decode("ascii")
    if not keychain_set(agent_id, seed_b64):
        return None
    return base64.b64encode(key.public_key().public_bytes_raw()).decode("ascii")


def public_key(agent_id: str) -> Optional[str]:
    """The stored public key for `agent_id` (base64), or None if absent."""
    key = _private_key_for_agent(agent_id)
    if key is None:
        return None
    return base64.b64encode(key.public_key().public_bytes_raw()).decode("ascii")


def public_key_from_seed(seed_b64: str) -> Optional[str]:
    """Public key (base64) derived from an explicit seed — tests/fixtures."""
    key = _private_key_from_seed(seed_b64)
    if key is None:
        return None
    return base64.b64encode(key.public_key().public_bytes_raw()).decode("ascii")


def list_ids() -> List[str]:
    """Agent-ids with signing keys present in the Keychain (sorted)."""
    dump = _run_security(["dump-keychain"])
    if dump is None:
        return []

    # Attributes are alphabetical, so "acct" precedes "svce" within an item.
    def quoted(line: str) -> Optional[str]:
        idx = line.rfind('="')
        if idx < 0 or not line.endswith('"'):
            return None
        return line[idx + 2:-1]

    ids = set()
    pending_acct: Optional[str] = None
    for line in dump.split("\n"):
        if '"acct"<blob>=' in line:
            pending_acct = quoted(line)
        if f'"svce"<blob>="{SERVICE}"' in line and pending_acct is not None:
            ids.add(pending_acct)
    return sorted(ids)


def remove(agent_id: str) -> bool:
    """Delete the signing key for `agent_id`."""
    return _run_security(["delete-generic-password", "-s", SERVICE, "-a", agent_id]) is not None


# MARK: - Sign / verify

def sign(data: bytes, agent_id: Optional[str] = None,
         seed_b64: Optional[str] = None) -> Optional[str]:
    """Sign `data` -> "ed25519:<base64 64-byte sig>".

    Pass `agent_id` to use the Keychain key, or `seed_b64` for an explicit
    32-byte seed (tests/fixtures)."""
    if seed_b64 is not None:
        key = _private_key_from_seed(seed_b64)
    elif agent_id is not None:
        key = _private_key_for_agent(agent_id)
    else:
        key = None
    if key is None:
        return None
    return SIG_PREFIX + base64.b64encode(key.sign(data)).decode("ascii")


def verify(data: bytes, sig: str, pubkey_b64: str) -> bool:
    """Verify an "ed25519:<base64>" signature over `data` with a base64 pubkey."""
    if not sig.startswith(SIG_PREFIX):
        return False
    sig_data = _b64decode(sig[len(SIG_PREFIX):])
    if sig_data is None or len(sig_data) != 64:
        return False
    pub_data = _b64decode(pubkey_b64)
    if pub_data is None or len(pub_data) != 32:
        return False
    try:
        pub = Ed25519PublicKey.from_public_bytes(pub_data)
        pub.verify(sig_data, data)
        return True
    except (InvalidSignature, ValueError):
        return False


# MARK: - Internals

def _b64decode(s: str) -> Optional[bytes]:
    try:
        return base64.b64decode(s, validate=True)
    except Exception:
        return None


def _private_key_from_seed(seed_b64: str) -> Optional[Ed25519PrivateKey]:
    seed = _b64decode(seed_b64)
    if seed is None or len(seed) != 32:
        return None
    try:
        return Ed25519PrivateKey.from_private_bytes(seed)
    except Exception:
        return None


def _private_key_for_agent(agent_id: str) -> Optional[Ed25519PrivateKey]:
    b64 = keychain_get(agent_id)
    if b64 is None:
        return None
    return _private_key_from_seed(b64)


# MARK: - Keychain (generic password via `security`, mirroring MeshKey)

def keychain_get(agent_id: str) -> Optional[str]:
    out = _run_security(["find-generic-password", "-s", SERVICE, "-a", agent_id, "-w"])
    if out is None:
        return None
    b64 = out.strip()
    return b64 if b64 else None


def keychain_set(agent_id: str, seed_b64: str) -> bool:
    # -U upserts (update if present)
    return _run_security(["add-generic-password", "-s", SERVICE, "-a", agent_id,
                          "-w", seed_b64, "-U", "-l", f"yx-sig-{agent_id}"]) is not None


def _run_security(args: List[str]) -> Optional[str]:
    try:
        r = subprocess.run(["/usr/bin/security"] + args,
                           capture_output=True, text=True, timeout=10)
    except Exception:
        return None
    if r.returncode != 0:
        return None
    return r.stdout
